src/dos/dos_memory.cpp

```diff
@@ -18,23 +18,90 @@
 
 #include "dosbox.h"
 
-#include "mem.h"
 #include "dos_inc.h"
+#include "enum.h"
+#include "mem.h"
 
-#define UMB_START_SEG 0x9fff
+enum class McbFaultStrategy { deny, repair, report, allow };
 
-static uint16_t memAllocStrategy = 0x00;
+// Constants
+// ~~~~~~~~~
 
-static void DOS_CompressMemory(void) {
-	uint16_t mcb_segment=dos.firstMCB;
+// MCB - DOS Memory Control Block Format
+// Offset Size   Description
+// 00     byte   'M' 4Dh  member of a MCB chain, (not last)
+//               'Z' 5Ah  indicates last entry in MCB chain
+// - the 'M' and 'Z' are said to represent Mark Zbikowski
+constexpr uint8_t mcb_type_M = {0x4d};
+constexpr uint8_t mcb_type_Z = {0x5a};
+
+// Upper member block starting segment
+constexpr uint16_t UMB_START_SEG = {0x9fff};
+
+static uint16_t memAllocStrategy = {0x00};
+
+constexpr uint16_t max_allowed_faults = {100};
+// not based on anything in particular. Player Manager 2 requires ~17 corrections.
+
+static auto mcb_fault_strategy = McbFaultStrategy::repair;
+
+void DOS_SetMcbFaultStrategy(const char *mcb_fault_strategy_pref)
+{
+	assert(mcb_fault_strategy_pref);
+	mcb_fault_strategy =
+	        enum_cast<McbFaultStrategy>(mcb_fault_strategy_pref).value();
+}
+
+// returns true if the MCB block needed triaging
+static bool triage_block(DOS_MCB &mcb, const uint8_t repair_type)
+{
+	auto mcb_type_is_valid = [&]() -> bool {
+		const auto t = mcb.GetType();
+		return t == mcb_type_M || t == mcb_type_Z;
+	};
+	if (mcb_type_is_valid())
+		return false;
+
+	switch (mcb_fault_strategy) {
+	case McbFaultStrategy::deny:
+		E_Exit("DOS_MEMORY: Exiting due corrupt MCB chain");
+		break;
+
+	case McbFaultStrategy::repair:
+		LOG_INFO("DOS_MEMORY: Repairing MCB block in segment %04xh from type '%02x' to '%02x'",
+		         mcb.GetPSPSeg(),
+		         mcb.GetType(),
+		         repair_type);
+		mcb.SetType(repair_type);
+		assert(mcb_type_is_valid());
+		break;
+
+	case McbFaultStrategy::report:
+		LOG_WARNING("DOS_MEMORY: Reporting MCB block in segment %04xh with corrupt type '%02x'",
+		            mcb.GetPSPSeg(),
+		            mcb.GetType());
+		break;
+
+	case McbFaultStrategy::allow:
+	default: break; // do nothing
+	}
```

Comment:
Why do we need a default block? I think the compiler does an exhaustiveness-check when dealing with enums in switch/case blocks.

Reply:
Even better; thanks (dropped it).